Add support for CLI XML import

jsachs opened the pull request:
## Type of change
- ✅ New feature (non-breaking change which adds functionality)

## Description and Context
The CLI now contains an "import" command that creates a new database from the specified XML export. The new database is in kdbx 4 format, and does not currently accept a keyfile in database creation.

This change is required to create new databases from XML backups.

Fixes #2458
 
## Testing strategy
CLI unit tests akin to the export tests

## Checklist:
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.
- ✅ I have added tests to cover my changes.

louib commented:
@jsachs I think there was a problem with the rebase and the newly merged YubiKey CLI option. See the build logs for details.

louib commented on src/core/Database.h:
```diff
@@ -73,6 +73,7 @@ class Database : public QObject
     bool save(QString* error = nullptr, bool atomic = true, bool backup = false);
     bool save(const QString& filePath, QString* error = nullptr, bool atomic = true, bool backup = false);
     bool extract(QByteArray&, QString* error = nullptr);
```
@jsachs We should rename that function `export` at that point I think.

jsachs replied:
I seem to remember a conversation somewhere about the naming of this function--if you think `export` is correct I can make the change as part of this PR.

louib replied:
I'd go with `export`, it makes sense especially now that we have `import`

jsachs replied:
Was this not already handled by #3278 ?

louib replied:
only the CLI command was renamed, not the `Database` function

jsachs replied:
https://en.cppreference.com/w/cpp/keyword/export

Do we want to find a way to override this, or just leave it as is?

droidmonkey replied:
If it is a function name it wouldn't clash with the keyword.

droidmonkey replied:
If it is a function name it wouldn't clash with the keyword.

louib replied:
mmmhhh not sure. We could also call them `exportToXML` and `importFromXML` to make it more explicit.

jsachs replied:
I think there is a clash based on the following error:
`keepassxc/src/core/Database.h:75:10: error: expected member name or ';' after declaration specifiers
    bool export(QByteArray&, QString* error = nullptr);`

jsachs replied:
I'd argue that based on the compiler complaints, it's fine to leave the underlying methods as `extract`, given that the user interface is going to be `import` and `export`.

droidmonkey replied:
Wow I take it back, apparently you cannot name a function "export" which sucks.

louib commented on src/cli/Import.cpp:
```diff
@@ -0,0 +1,121 @@
+ * @return Pointer to the PasswordKey or null if passwordkey is skipped
+ *         by user
+ */
+QSharedPointer<PasswordKey> Import::getPasswordFromStdin()
```
@jsachs Is that not defined in `Create` too? Maybe it should go into `cli::Utils`

jsachs replied:
Yep, seems correct. I'll move it over.

louib commented on tests/TestCli.cpp:
```diff
@@ -175,7 +187,7 @@ QSharedPointer<Database> TestCli::readTestDatabase() const
 void TestCli::testCommand()
 {
-    QCOMPARE(Command::getCommands().size(), 17);
+    QCOMPARE(Command::getCommands().size(), 15);
```
@jsachs 18?

jsachs replied:
Yep, this was another upstream mistake.

jsachs commented:
@droidmonkey any idea what is causing the Ubuntu build issues? It seems to be manifesting in other feature PRs as well.

droidmonkey commented:
Weird, looks like the CI cannot open a temporary file.

louib commented on tests/TestCli.cpp:
```diff
@@ -944,6 +957,56 @@ void TestCli::testGenerate()
+
+    Utils::Test::setNextPassword("a");
+    auto db = QSharedPointer<Database>(Utils::unlockDatabase(databaseFilename, true, "", "", Utils::DEVNULL));
+    QVERIFY(db);
```
@jsachs is there an entry in the xml database? we could add a basic assert here on the presence of the entry.

louib reviewed (commented):
This one will need a rebase now that the `open` command's PR was merged

louib commented on tests/TestCli.cpp:
```diff
@@ -983,6 +996,59 @@ void TestCli::testGenerate()
+    pos = m_stdoutFile->pos();
+    Utils::Test::setNextPassword("a");
+    importCmd.execute({"import", "-q", m_xmlFile->fileName(), databaseFilenameQuiet});
+    m_stdoutFile->seek(pos);
```
@jsachs here we're not actually testing that there was no output on stdout with the quiet option. And it would probably fail since it doesn't seem to be implemented.

Should be fairly easy, you just have to defined the output stream as
```
TextStream outputTextStream(parser->isSet(Command::QuietOption) ? Utils::DEVNULL : Utils::STDOUT, QIODevice::WriteOnly);
```

louib commented:
@jsachs for some reason the tests are still failing
```
Actual   (m_stdoutFile->readAll())                                                           : "Insert password to encrypt database (Press enter to leave blank): \nSuccessfully imported database.\n"
Expected (QByteArray("Insert password to encrypt database (Press enter to leave blank): \n")): "Insert password to encrypt database (Press enter to leave blank): \n"
Loc: [/Users/kpxc/keepassxc/38351/tests/TestCli.cpp(1048)]
```
:interrobang: 

I'd approve this one otherwise

jsachs commented:
Any idea why the `-q` option would fail in a test? I can confirm that this works correctly when running the CLI manually.

louib commented on src/cli/Import.h:
```diff
@@ -0,0 +1,32 @@
+
+#include "Command.h"
+
+#include "keys/PasswordKey.h"
```
@jsachs I don't think you need that import anymore.

louib commented:
@jsachs I have no clue tbh :grimacing:

louib reviewed (approved):
@jsachs thanks for the PR!
@droidmonkey all yours

jsachs committed:
CLI: Add Import XML command (#3572)

The CLI now contains an "import" command that creates a new database from the specified XML export. The new database is in kdbx 4 format, and does not currently accept a keyfile in database creation.

This change is required to create new databases from XML backups.

Fixes #2458

droidmonkey commented:
Squashed all the commits and bumped up the default KDF to Argon2 to save as KDBX4
